clang++ 6.0.0 with libc++ fails to compile call.cc

```
[ 19%] Building CXX object lib/evaluate/CMakeFiles/FortranEvaluate.dir/call.cc.o
In file included from /home/ryuta/packages/llvm/f18/lib/evaluate/call.cc:15:
In file included from /home/ryuta/packages/llvm/f18/lib/evaluate/call.h:18:
In file included from /home/ryuta/packages/llvm/f18/lib/evaluate/common.h:18:
In file included from /home/ryuta/packages/llvm/f18/lib/evaluate/intrinsics-library.h:26:
In file included from /usr/local/bin/../include/c++/v1/functional:484:
/usr/local/bin/../include/c++/v1/type_traits:3141:38: error: implicit instantiation of undefined template 'Fortran::evaluate::Expr<Fortran::evaluate::SomeType>'
    : public integral_constant<bool, __is_constructible(_Tp, _Args...)>
                                     ^
/usr/local/bin/../include/c++/v1/type_traits:3380:7: note: in instantiation of template class
      'std::__1::is_constructible<Fortran::common::Indirection<Fortran::evaluate::Expr<Fortran::evaluate::SomeType>, true>, Fortran::evaluate::Expr<Fortran::evaluate::SomeType> >' requested here
    = is_constructible<_Tp, _Args...>::value;
      ^
/usr/local/bin/../include/c++/v1/variant:1152:19: note: in instantiation of variable template specialization
      'std::__1::is_constructible_v<Fortran::common::Indirection<Fortran::evaluate::Expr<Fortran::evaluate::SomeType>, true>, Fortran::evaluate::Expr<Fortran::evaluate::SomeType> >' requested here
      enable_if_t<is_constructible_v<_Tp, _Arg>, int> = 0>
                  ^
/usr/local/bin/../include/c++/v1/variant:1154:13: note: while substituting prior template arguments into non-type template parameter [with _Arg =
      Fortran::evaluate::Expr<Fortran::evaluate::SomeType>, $1 = 0, $2 = 0, $3 = 0, _Tp = Fortran::common::Indirection<Fortran::evaluate::Expr<Fortran::evaluate::SomeType>, true>, _Ip = 0]
  constexpr variant(_Arg&& __arg) noexcept(
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/ryuta/packages/llvm/f18/lib/evaluate/call.h:62:52: note: while substituting deduced template arguments into function template 'variant' [with _Arg =
      Fortran::evaluate::Expr<Fortran::evaluate::SomeType>, $1 = (no value), $2 = (no value), $3 = (no value), _Tp = (no value), _Ip = (no value), $6 = (no value)]
  explicit ActualArgument(Expr<SomeType> &&x) : u_{std::move(x)} {}
                                                   ^
/home/ryuta/packages/llvm/f18/lib/evaluate/common.h:199:28: note: template is declared here
template<typename A> class Expr;
                           ^
In file included from /home/ryuta/packages/llvm/f18/lib/evaluate/call.cc:15:
/home/ryuta/packages/llvm/f18/lib/evaluate/call.h:62:49: error: no matching constructor for initialization of 'std::variant<common::CopyableIndirection<Expr<SomeType> >, AssumedType>'
      (aka 'variant<Fortran::common::Indirection<Fortran::evaluate::Expr<Fortran::evaluate::SomeType>, true>, Fortran::evaluate::ActualArgument::AssumedType>')
  explicit ActualArgument(Expr<SomeType> &&x) : u_{std::move(x)} {}
                                                ^ ~~~~~~~~~~~~~~
/usr/local/bin/../include/c++/v1/variant:1141:3: note: candidate constructor not viable: cannot convert argument of incomplete type 'typename remove_reference<Expr<SomeType> &>::type'
      (aka 'Fortran::evaluate::Expr<Fortran::evaluate::SomeType>') to 'const std::__1::variant<Fortran::common::Indirection<Fortran::evaluate::Expr<Fortran::evaluate::SomeType>, true>,
      Fortran::evaluate::ActualArgument::AssumedType>' for 1st argument
  variant(const variant&) = default;
  ^
/usr/local/bin/../include/c++/v1/variant:1142:3: note: candidate constructor not viable: cannot convert argument of incomplete type 'typename remove_reference<Expr<SomeType> &>::type'
      (aka 'Fortran::evaluate::Expr<Fortran::evaluate::SomeType>') to 'std::__1::variant<Fortran::common::Indirection<Fortran::evaluate::Expr<Fortran::evaluate::SomeType>, true>,
      Fortran::evaluate::ActualArgument::AssumedType>' for 1st argument
  variant(variant&&) = default;
  ^
/usr/local/bin/../include/c++/v1/variant:1154:13: note: candidate template ignored: requirement
      'is_constructible_v<Fortran::common::Indirection<Fortran::evaluate::Expr<Fortran::evaluate::SomeType>, true>, Fortran::evaluate::Expr<Fortran::evaluate::SomeType> >' was not satisfied [with _Arg =
      Fortran::evaluate::Expr<Fortran::evaluate::SomeType>, $1 = 0, $2 = 0, $3 = 0, _Tp = Fortran::common::Indirection<Fortran::evaluate::Expr<Fortran::evaluate::SomeType>, true>, _Ip = 0]
  constexpr variant(_Arg&& __arg) noexcept(
            ^
/usr/local/bin/../include/c++/v1/variant:1163:22: note: candidate template ignored: could not match 'in_place_index_t' against 'Expr'
  explicit constexpr variant(
                     ^
/usr/local/bin/../include/c++/v1/variant:1191:22: note: candidate template ignored: could not match 'in_place_type_t' against 'Expr'
  explicit constexpr variant(in_place_type_t<_Tp>, _Args&&... __args) noexcept(
                     ^
/usr/local/bin/../include/c++/v1/variant:1138:13: note: candidate constructor template not viable: requires 0 arguments, but 1 was provided
  constexpr variant() noexcept(is_nothrow_default_constructible_v<__first_type>)
            ^
/usr/local/bin/../include/c++/v1/variant:1177:22: note: candidate constructor template not viable: requires at least 2 arguments, but 1 was provided
  explicit constexpr variant(
                     ^
/usr/local/bin/../include/c++/v1/variant:1204:22: note: candidate constructor template not viable: requires at least 2 arguments, but 1 was provided
  explicit constexpr variant(
                     ^
/home/ryuta/packages/llvm/f18/lib/evaluate/call.cc:33:6: error: no viable overloaded '='
  u_ = std::move(expr);
  ~~ ^ ~~~~~~~~~~~~~~~
/usr/local/bin/../include/c++/v1/variant:1213:12: note: candidate function not viable: no known conversion from 'typename remove_reference<Expr<SomeType> &>::type'
      (aka 'Fortran::evaluate::Expr<Fortran::evaluate::SomeType>') to 'const std::__1::variant<Fortran::common::Indirection<Fortran::evaluate::Expr<Fortran::evaluate::SomeType>, true>,
      Fortran::evaluate::ActualArgument::AssumedType>' for 1st argument
  variant& operator=(const variant&) = default;
           ^
/usr/local/bin/../include/c++/v1/variant:1214:12: note: candidate function not viable: no known conversion from 'typename remove_reference<Expr<SomeType> &>::type'
      (aka 'Fortran::evaluate::Expr<Fortran::evaluate::SomeType>') to 'std::__1::variant<Fortran::common::Indirection<Fortran::evaluate::Expr<Fortran::evaluate::SomeType>, true>,
      Fortran::evaluate::ActualArgument::AssumedType>' for 1st argument
  variant& operator=(variant&&) = default;
           ^
/usr/local/bin/../include/c++/v1/variant:1225:12: note: candidate template ignored: requirement
      'is_constructible_v<Fortran::common::Indirection<Fortran::evaluate::Expr<Fortran::evaluate::SomeType>, true>, Fortran::evaluate::Expr<Fortran::evaluate::SomeType> >' was not satisfied [with _Arg =
      Fortran::evaluate::Expr<Fortran::evaluate::SomeType>, $1 = 0, _Tp = Fortran::common::Indirection<Fortran::evaluate::Expr<Fortran::evaluate::SomeType>, true>, _Ip = 0]
  variant& operator=(_Arg&& __arg) noexcept(
           ^
3 errors generated.
make[2]: *** [lib/evaluate/CMakeFiles/FortranEvaluate.dir/call.cc.o] Error 1
make[1]: *** [lib/evaluate/CMakeFiles/FortranEvaluate.dir/all] Error 2
make: *** [all] Error 2
```

--- a/lib/evaluate/call.cc
+++ b/lib/evaluate/call.cc
@@ -20,6 +20,12 @@
 
 namespace Fortran::evaluate {
 
+ActualArgument::ActualArgument(Expr<SomeType> &&x) : u_{std::move(x)} {}
+ActualArgument::ActualArgument(common::CopyableIndirection<Expr<SomeType>> &&v)
+  : u_{std::move(v)} {}
+ActualArgument::ActualArgument(AssumedType x) : u_{x} {}
+ActualArgument::~ActualArgument() {}
+
 ActualArgument::AssumedType::AssumedType(const semantics::Symbol &symbol)
   : symbol_{&symbol} {
   const semantics::DeclTypeSpec *type{symbol.GetType()};

--- a/lib/evaluate/call.h
+++ b/lib/evaluate/call.h
@@ -59,11 +59,10 @@ public:
     const semantics::Symbol *symbol_;
   };
 
-  explicit ActualArgument(Expr<SomeType> &&x) : u_{std::move(x)} {}
-  explicit ActualArgument(common::CopyableIndirection<Expr<SomeType>> &&v)
-    : u_{std::move(v)} {}
-  explicit ActualArgument(AssumedType x) : u_{x} {}
-
+  explicit ActualArgument(Expr<SomeType> &&);
+  explicit ActualArgument(common::CopyableIndirection<Expr<SomeType>> &&);
+  explicit ActualArgument(AssumedType);
+  ~ActualArgument();
   ActualArgument &operator=(Expr<SomeType> &&);
 
   Expr<SomeType> *GetExpr() {

--- a/lib/evaluate/characteristics.cc
+++ b/lib/evaluate/characteristics.cc
@@ -187,6 +187,9 @@ bool IsOptional(const DummyArgument &da) {
       da);
 }
 
+FunctionResult::FunctionResult(DynamicType t) : u{TypeAndShape{t}} {}
+FunctionResult::FunctionResult(TypeAndShape &&t) : u{std::move(t)} {}
+FunctionResult::FunctionResult(Procedure &&p) : u{std::move(p)} {}
 FunctionResult::~FunctionResult() = default;
 
 bool FunctionResult::operator==(const FunctionResult &that) const {

--- a/lib/evaluate/characteristics.h
+++ b/lib/evaluate/characteristics.h
@@ -127,9 +127,9 @@ std::optional<DummyArgument> CharacterizeDummyArgument(
 struct FunctionResult {
   ENUM_CLASS(Attr, Allocatable, Pointer, Contiguous)
   DECLARE_CONSTRUCTORS_AND_ASSIGNMENTS(FunctionResult)
-  explicit FunctionResult(DynamicType t) : u{TypeAndShape{t}} {}
-  explicit FunctionResult(TypeAndShape &&t) : u{std::move(t)} {}
-  explicit FunctionResult(Procedure &&p) : u{std::move(p)} {}
+  explicit FunctionResult(DynamicType);
+  explicit FunctionResult(TypeAndShape &&);
+  explicit FunctionResult(Procedure &&);
   ~FunctionResult();
   bool operator==(const FunctionResult &) const;
   static std::optional<FunctionResult> Characterize(

